### PR TITLE
FOCUS #154: Create Charge Frequency Dimension

### PR DESCRIPTION
**Pull request by SanjnaSrivatsaVMware**

### CODE REVIEW

File: specification/dimensions/chargefrequency.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Charge Frequency
+
+Charge Frequency indicates how often a charge would occur - periodically, usage-based, or one-time charge. When the charge frequency is periodic (typically monthly), it can be inferred that this type of charge can be seen in past and future bill periods (within constraints like charge type or possible expiry). Such records indicate charges for the acquisition of a service or resource bought upfront or on a recurring basis.
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:3]: A few thoughts:
1. I believe we try to avoid listing the values in the description.
2. The "inferred" comment feels odd here. The intent is spot on, but this may be the first or last charge. Curious what others think.
3. "Billing periods" (not "bill periods").
4. Should we link the Charge Frequency to Charge Period Start/End? How do those relate? Can we use that to know when the charge will recur?

File: specification/dimensions/chargefrequency.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+| Column required | True                                     |
+| Data type       | String                                   |
+| Allows nulls    | False                                    |
+| Value format    | list-of-values                           |
+
+Allowed values:
+
+| Value       | Description                                                                                                                                                                                         |
+| ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Periodic    | This charge type will occur on a periodic cadence which can be daily, hourly or monthly which indicate charges for the acquisition of a service or resource bought upfront or on a recurring basis. |
```
[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:36]: Thank you @flowerinthenight, this is good referrence. Given that we are trying to indicate that this charge type shows up in every bill period, should we use periodic?

File: specification/dimensions/chargefrequency.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Charge Frequency
+
+Charge Frequency indicates how often a charge would occur - periodically, usage-based, or one-time charge. When the charge frequency is periodic (typically monthly), it can be inferred that this type of charge can be seen in past and future bill periods (within constraints like charge type or possible expiry). Such records indicate charges for the acquisition of a service or resource bought upfront or on a recurring basis.
+When charge frequency is one-time, this can be an indicator of a prepaid purchase which cannot be directly tied to usage in that billing period. Usage-based charges will occur as frequently as that type of usage exists.
```
[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:4]: I don't think I quite understood this comment

File: specification/dimensions/chargefrequency.md (hunk 1 above)
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:3]: 1. I have turned it into a few example values
2. Removed inferred
3. changed 
4. I don't have a strong opinion either way. I do see the benefit of having a start and end to understand when the charge will recur or until when, but I know we want to be really discriminate about adding more dimensions

Comment (flanakin) [reply specification/dimensions/chargefrequency.md]: ChargePeriodStart/End already exist :-) I was referring to the ones we added in v0.5.

File: specification/dimensions/chargefrequency.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (flanakin) [reply specification/dimensions/chargefrequency.md:4]: When I read this, it feels like we're focusing more on describing the values than the column. I'm thinking the description of the values should probably be in the "allowed values" section below.

Also curious what others think. @udam-f2?

**Review by jasonkellyAP — APPROVED** [review]

File: specification/dimensions/chargefrequency.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (caseydoran) [reply specification/dimensions/chargefrequency.md:36]: **Periodic** makes sense to me given the above, although **recurring** is more often used in this context. 
examples:
https://docs.aws.amazon.com/cur/latest/userguide/regular-reserved-instances.html#recurring-monthly
https://learn.microsoft.com/en-us/azure/cost-management-billing/reservations/prepare-buy-reservation
https://cloud.google.com/docs/cuds-spend-based

based on expected language in context, I'd suggest **recurring** as the term we use

**Review by caseydoran — CHANGES_REQUESTED** [review]
I prefer updating "periodic" to "recurring" based on that being the language the 3 major CSPs currently use in this context.

File: specification/dimensions/chargefrequency.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rodneyjoyce) [reply specification/dimensions/chargefrequency.md:4]: Based on what we did for 0.5 fields, I think we should keep the description simple and let the Allowed Values speak for themselves. "Charge Frequency indicates how often a charge would occur." I think this is sufficient IMO

File: specification/dimensions/chargefrequency.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodneyjoyce) [reply specification/dimensions/chargefrequency.md:36]: I prefer **recurring**

File: specification/dimensions/chargefrequency.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+| Allows nulls    | False                                    |
+| Value format    | list-of-values                           |
+
+Allowed values:
+
+| Value       | Description                                                                                                                                                                                         |
+| ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Periodic    | This charge type will occur on a periodic cadence which can be daily, hourly or monthly which indicate charges for the acquisition of a service or resource bought upfront or on a recurring basis. |
+| Usage Based | Charges based on the quantity and frequency of a service or resource that was consumed in a given period of time.                                                                                    |
+| One time    | Charges that will occur only and indicate a one-time purchase of a contract/commitment or pre-paid service.
```
Comment (rodneyjoyce) [inline specification/dimensions/chargefrequency.md:38]: Should this not be "One-time"? If not, then One Time?

File: specification/dimensions/chargefrequency.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:34]: Recurring it is!

File: specification/dimensions/chargefrequency.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:4]: I might have gotten sidetracked looking at the style of writing from another Dimension. I have kept the Description short and precise and focused all value related information to the 'allowed-values' section.

File: specification/dimensions/chargefrequency.md (hunk 1 above)
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md]: In this case, the charge period start and end refer to the month-year that the recurring charge has started and when that charge will expire or stop showing up, correct?

**Review by joe-ferrero — COMMENTED** [review]
Looks good to me.  I like the call out for one-time payments - it can either be a pre-payment of usage charges or a truly one-time fee - such as a setup fee or a service fee.

File: specification/dimensions/chargefrequency.md
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:1]: Don't forget to add this to `dimensions.mdpp` in the same folder.

File: specification/dimensions/chargefrequency.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Charge Frequency
+
+Charge Frequency indicates how often a charge would occur.
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:3]: Noticed you use `will` in other areas.
Also add scenarios:
```suggestion
Charge Frequency indicates how often a charge will occur. Charge Frequency is commonly used for scenarios like forecasting and differentiating between one-time and recurring fees for commitment-based discounts.
```
(There might be a better one to call out than CBD here.)

File: specification/dimensions/chargefrequency.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# Charge Frequency
+
+Charge Frequency indicates how often a charge would occur. 
+
+The ChargeFrequency column MUST be present and MUST NOT be null or empty. This column is of type String and MUST be one of the
+allowed values.
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:6]: nit: Keep on the same line
```suggestion
The ChargeFrequency column MUST be present and MUST NOT be null or empty. This column is of type String and MUST be one of the allowed values.
```

File: specification/dimensions/chargefrequency.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+| Data type       | String                                   |
+| Allows nulls    | False                                    |
+| Value format    | list-of-values                           |
+
+Allowed values:
+
+| Value       | Description                                                                                                                                                                                         |
+| ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Recurring   | This charge type will occur on a periodic cadence which can be daily, hourly or monthly which indicate charges for the acquisition of a service or resource bought upfront or on a recurring basis. |
+| Usage Based | Charges based on the quantity and frequency of a service or resource that was consumed in a given period of time.                                                                                    |
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:34]: I believe this should be hyphenated, since it's still describing a charge.
```suggestion
| Usage-Based | Charges based on the quantity and frequency of a service or resource that was consumed in a given period of time.                                                                                    |
```

File: specification/dimensions/chargefrequency.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+| One-time    | Charges that will occur only and indicate a one-time purchase of a contract/commitment or pre-paid service.
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:35]: I believe this should be a capital T for consistency with our title casing.
```suggestion
| One-Time    | Charges that will occur only and indicate a one-time purchase of a contract/commitment or pre-paid service.     
```

File: specification/dimensions/chargefrequency.md (hunk 5 above)
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:3]: Would it make sense to add a reference to ChargePeriodStart/End here? Maybe something like this...?

> _While Charge Frequency indicates whether a charge will reoccur, Charge Period Start and Charge Period End specify the period for which this charge applies. This can typically be used to also understand the recurrence period (e.g., monthly, yearly)._

File: supporting_content/dimensions/chargefrequency.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+## Example usage scenarios
+
+Current values observed in billing data for various scenarios:
+
+| Provider  | Data set                 | Example Value                |
+| --------- | ------------------------ | ---------------------------- |
+| AWS       | CUR                      | Anniversary,Purchase,Refund  |
+| GCP       | Big Query Billing Export | NA                           |
+| Microsoft | Cost details             | Recurring,Monthly,UsageBased |
```
Comment (flanakin) [inline supporting_content/dimensions/chargefrequency.md:22]: I didn't think we had "Monthly", but I do know that was the original intent, so I wouldn't be surprised if someone snuck it in.
```suggestion
| Microsoft | Cost details             | OneTime, Recurring, Monthly, UsageBased |
```

File: supporting_content/dimensions/chargefrequency.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+
+
+## Reference
+- Microsoft: [understand-usage-details-fields](https://learn.microsoft.com/en-us/azure/cost-management-billing/automate/understand-usage-details-fields)
+- AWS: [understand-bill-details-fields](https://docs.aws.amazon.com/cur/latest/userguide/billing-columns.html)
+
+## Discussion / Scratch space
```
Comment (flanakin) [inline supporting_content/dimensions/chargefrequency.md:30]: Can you add details about the recurring/periodic discussion?

**Review by flanakin — APPROVED** [review]

**Review by rodneyjoyce — COMMENTED** [review]
I have reviewed but have a question:

"Usage-based" is technically not a Frequency. Is this here because we don't want to use NULL if the cost record does not have a Frequency?

**Comment by flanakin** [comment]
> I have reviewed but have a question:
> 
> "Usage-based" is technically not a Frequency. Is this here because we don't want to use NULL if the cost record does not have a Frequency?

It's supposed to mean "charges will recur based on your usage". Do you think null implies that?

File: specification/dimensions/chargefrequency.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+| Data type       | String                                   |
+| Allows nulls    | False                                    |
+| Value format    | list-of-values                           |
+
+Allowed values:
+
+| Value       | Description                                                                                                                                                                                         |
+| ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Recurring   | This charge type will occur on a periodic cadence which can be daily, hourly or monthly which indicate charges for the acquisition of a service or resource bought upfront or on a recurring basis. |
+| Usage-Based | Charges based on the quantity and frequency of a service or resource that was consumed in a given period of time.                                                                                    |
```
[the remark this answers is not in the document]
Comment (hekike) [reply specification/dimensions/chargefrequency.md:35]: I personally find it confusing as usage-based can also be charged monthly.
Stripe, for [example](https://stripe.com/docs/api/subscriptions/object#subscription_object-items-data-price-recurring), solved it as:

```json
{
  ...
   "recurring": {
            "aggregate_usage": "sum",
            "interval": "month",
            "interval_count": 1,
            "usage_type": "metered"
         }
```

[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:33]: Yes Udam, the conversation happened here https://finopsfoundation.slack.com/archives/C05JTFK4ZT2/p1694647300587519. 
TLDR: 
The way I see it is that logically it does make sense to have our current 3 buckets of frequencies -
One indicates that certain entries are recurring - meaning they will re-occur every so often irrespective of utilization or other aspects- indicating frequency
Another indicates that this charge is one-time - no recurrence - again indicating frequency
The last one is also recurring (usage-based) - meaning that this charge MAY reoccur but depending on some constraints the service continues to be consumed - IMO this also indicates frequency like the first one, but maybe the term we use here can be changed if anyone feels strongly.

[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:33]: @hekike We can change the value name from usage-based to something else but what we are trying to convey is that usage-based charges will occur only as frequently as the usage occurs. This seems like an implicit thing to say but it implies the occurrence is a relationship to something (in this case usage). 
If you used SQS this month, your charges will be usage-based which means you used it, hence you see these charges this month. If you do not use it next month, these charges will no longer occur.

[the remark this answers is not in the document]
Comment (SanjnaSrivatsaVMware) [reply specification/dimensions/chargefrequency.md:33]: @hekike We can change the value name from usage-based to something else but what we are trying to convey is that usage-based charges will occur only as frequently as the usage occurs. This seems like an implicit thing to say but it implies the occurrence is a relationship to something (in this case usage). 
If you used SQS this month, your charges will be usage-based which means you used it, hence you see these charges this month. If you do not use it next month, these charges will no longer occur.

**Review by ljadvey — APPROVED** [review]

File: specification/dimensions/chargefrequency.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# Charge Frequency
+
+Charge Frequency indicates how often a charge will occur. Along with the Charge Period column, the Charge Frequency can be used to understand recurrence periods (e.g., monthly, yearly), forecast upcoming charges, and differentiate between one-time and recurring fees for purchases.
+
+The ChargeFrequency column MUST be present and MUST NOT be null or empty. This column is of type String and MUST be one of the allowed values.
```
Comment (marcin-walkow) [inline specification/dimensions/chargefrequency.md:5]: A minor tweak to keep the standard and remove redundancy:
The ChargeFrequency column MUST be present **in the billing data** and MUST NOT **be null**.

Comment (marcin-walkow) [inline specification/dimensions/chargefrequency.md:5]: A minor tweak to keep the standard and remove redundancy:
The ChargeFrequency column MUST be present **in the billing data** and MUST NOT **be null**.

File: specification/dimensions/chargefrequency.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (aqu-erp) [reply specification/dimensions/chargefrequency.md:33]: > - Another indicates that this charge is one-time - no recurrence - again indicating frequency
> - The last one is also recurring (usage-based) - meaning that this charge MAY reoccur but depending on some constraints the service continues to be consumed [usage-based charges will occur only as frequently as the usage occurs]

Both options describe a charge occurring only at the point of consumption. I think what's being conflated are the characteristics of the chargeType. An upfront purchase is non-recurring, just as usage is non-recurring. Both are not charged again until another instance of consumption has occurred. 

What kind of consumption would propagate `chargeType.usage` and `chargeFrequency.one-time`?

To draw it out, from a charge and billing POV, an upfront RI is a cost incurred in one billing period, similar to EC2 usage. Neither will recur into another billing period. Another separate and distinct upfront RI or record of EC2 usage can be charged in the next period, if they occur.

**Review by aqu-erp — CHANGES_REQUESTED** [review]

File: specification/dimensions/chargefrequency.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (udam-f2) [reply specification/dimensions/chargefrequency.md:33]: Seems like something that needs a bit of discussion and hopefully we can get consensus? @SanjnaSrivatsaVMware are you able to drive consensus with the folks that have expressed their views here? Thanks!

[the remark this answers is not in the document]
Comment (hekike) [reply specification/dimensions/chargefrequency.md:33]: Thanks @SanjnaSrivatsaVMware, I'm okay with naming especially with clarifying when the charge occurs as @aqu-erp recommended.

File: specification/dimensions/chargefrequency.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
+# Charge Frequency
+
+Charge Frequency indicates how often a charge will occur. Charge Frequency is commonly used for scenarios like forecasting and differentiating between one-time and recurring fees for commitment-based discounts.
```
[the remark this answers is not in the document]
Comment (flanakin) [reply specification/dimensions/chargefrequency.md]: @SanjnaSrivatsaVMware, can you make sure all of these questions are answered in the allowed values descriptions? As long as they are, we should be good. If anything isn't clear, then that's probably a hint that we haven't quite hit the mark.

File: specification/dimensions/chargefrequency.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (flanakin) [reply specification/dimensions/chargefrequency.md:33]: Fwiw, "Usage-Based" is technically accurate:
"How frequently will I get charged?"
"It's based on your usage – When you use the service, you'll be charged."

I like @hekike's "usage-based can be charged monthly" scenario. That's pointing out something we haven't really discussed in detail, which is that there may be "Usage" charges that are billed monthly, which would be a "Recurring" frequency. We should be more explicit and say that "Usage-Based" means that you'll be charged when usage occurs. If there's a charge that happens monthly, even if it's based on usage, then that's not a "usage-based frequency". That's a recurring frequency.

I also like @aqu-erp's description.

@SanjnaSrivatsaVMware, can you get both of those into the supplemental notes? Seems like we're good here.

File: specification/dimensions/chargefrequency.md [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+| ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| Recurring   | This charge type will occur on a periodic cadence which can be daily, hourly or monthly which indicate charges for the acquisition of a service or resource bought upfront irrespective of their utilization on a recurring basis.               |
+| Usage-Based | Charges based on the quantity and frequency of a service or resource that was consumed in a given period of time. If the consumption/utilization of the service or resource stops, these charges will no longer be incurred.                     |
+| One-Time    | Charges that will occur only once and indicate a one-time purchase of a contract/commitment or pre-paid service. In other words, it can either be a pre-payment of usage charges or a truly one-time fee - such as a setup fee or a service fee. |
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:34]: Reworded this for consistency and to focus on when the charge happens with expectations around the cadence. I removed the explicit examples since we have some in supporting content and they didn't cover all possible charges. I moved one-time first thinking that might flow more smoothly (once → multiple times → when used).
```suggestion
| Value       | Description                                                                                                                                                                                                                                                                         |
| ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| One-Time    | Charges that only happen once and will not repeat. One-time charges are typically recorded on the hour or day when the cost was incurred.                                                                                                                                           |
| Recurring   | Charges that repeat on a periodic cadence (e.g., weekly, monthly) regardless of whether the product or service was used. Recurring charges typically happen on the same day or point within every period. The charge date does not change based on how or when the service is used. |
| Usage-Based | Charges that repeat every time the service is used. Usage-based charges are typically recorded hourly or daily, based on the granularity of the cost data, but only when the service is used. Usage-based charges are not recorded when the service is not used.                    |
```

**Review by flanakin — APPROVED** [review]

**Review by udam-f2 — APPROVED** [review]
Minor suggestions. Accept the rewording below if it makes sense.

File: specification/dimensions/chargefrequency.md [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+| ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| One-Time    | Charges that only happen once and will not repeat. One-time charges are typically recorded on the hour or day when the cost was incurred.                                                                                                                                           |
+| Recurring   | Charges that repeat on a periodic cadence (e.g., weekly, monthly) regardless of whether the product or service was used. Recurring charges typically happen on the same day or point within every period. The charge date does not change based on how or when the service is used. |
+| Usage-Based | Charges that repeat every time the service is used. Usage-based charges are typically recorded hourly or daily, based on the granularity of the cost data, but only when the service is used. Usage-based charges are not recorded when the service is not used.                    |
```
[the remark this answers is not in the document]
Comment (flanakin) [reply specification/dimensions/chargefrequency.md:34]:
```suggestion
| Usage-Based | Charges that repeat every time the service is used. Usage-based charges are typically recorded hourly or daily, based on the granularity of the cost data but only when the service is used. Usage-based charges are not recorded when the service is not used.                    |
```

File: specification/dimensions/chargefrequency.md [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+| One-Time    | Charges that only happen once and will not repeat. One-time charges are typically recorded on the hour or day when the cost was incurred.                                                                                                                                           |
+| Recurring   | Charges that repeat on a periodic cadence (e.g., weekly, monthly) regardless of whether the product or service was used. Recurring charges typically happen on the same day or point within every period. The charge date does not change based on how or when the service is used. |
+| Usage-Based | Charges that repeat every time the service is used. Usage-based charges are typically recorded hourly or daily, based on the granularity of the cost data, for the period when the service was used (referred to as Charge Period). Usage-based charges are not recorded when the service is not used.                    |
```
Comment (flanakin) [inline specification/dimensions/chargefrequency.md:34]:
```suggestion
| Usage-Based | Charges that repeat every time the service is used. Usage-based charges are typically recorded hourly or daily, based on the granularity of the cost data for the period when the service was used (referred to as Charge Period). Usage-based charges are not recorded when the service is not used.                    |
```

**Review by flanakin — APPROVED** [review]

**Comment by rupagcp** [comment]
I'm not sure "recurring" semantically makes sense for Tax, if the definition is "Charges that repeat on a periodic cadence regardless of whether the product or service was used". If products were not used, then there is no Tax for a given Invoice. As in, the tax does not itself "recur".
